### PR TITLE
Create a custom type for each java class used in python.

### PR DESCRIPTION
**Pull request by bsteffensmeier**

### CODE REVIEW

File: src/test/python/test_lists.py [hunk 1]
```diff
@@ -88,7 +89,7 @@ def test_setslice(self):
     def test_add(self):
         jlist = makeJavaList()
         x = jlist + [1, 2, 3]
-        self.assertIn('jep.PyJList', str(type(x)))
+        self.assertTrue(isinstance(x, List.__pytype__))
```
Comment (ndjensen) [inline src/test/python/test_lists.py:92]: What happens if you replace `List.__pytype__` with `ArrayList`?

**Review by ndjensen — APPROVED** [review]
I would like to discuss this a little more in depth offline.  I don't quite understand how it all works.

**Comment by ndjensen** [comment]
Do we need to remove Python 3.3 and 3.4 support?  AppVeyor failed with those.

File: src/main/c/Jep/pyembed.c [hunk 2]
```diff
@@ -228,15 +228,16 @@ void pyembed_preinit(JNIEnv *env,
  *
  * Furthermore, we need to ensure that the inheritance tree is built in the
  * correct order, i.e. from the top down.  For example, we need to set that
- * PyJIterable's tp_base extends PyJObject before we set that PyJCollection's
- * tp_base extends PyJIterable.
+ * PyJCollection's tp_base extends PyJIterbale before we set that PyJList's
```
[the remark this answers is not in the document]
Comment (bsteffensmeier) [reply src/main/c/Jep/pyembed.c]: Done

File: src/main/c/Objects/pyjtype.c [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+        PyObject* superType = (PyObject*) pyjtype_get_cached(env, fqnToPyType, super);
+        (*env)->DeleteLocalRef(env, super);
+        if (!superType) {
+            Py_DECREF(bases);
+            return NULL;
+        }
+        PyTuple_SET_ITEM(bases, baseIdx, superType);
+        baseIdx += 1;
+    }
+    jint interfacesIdx = 0;
```
[the remark this answers is not in the document]
Comment (bsteffensmeier) [reply src/main/c/Objects/pyjtype.c:136]: Done

File: src/main/c/Objects/pyjtype.c [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+    Py_XDECREF(shortName);
+    if (type) {
+        PyDict_SetItem(fqnToPyType, typeName, (PyObject*) type);
+    }
+    return type;
+}
+
+/*
+ * Either create a new type for the class or return a cached type if a type
+ * has already been provided. This function was seperated from PyJType_Get
```
[the remark this answers is not in the document]
Comment (bsteffensmeier) [reply src/main/c/Objects/pyjtype.c]: Done

File: src/test/python/test_lists.py (hunk 1 above)
Comment (bsteffensmeier) [reply src/test/python/test_lists.py:92]: ArrayList is a pyjclass, so it would throw an error. ArrayList.__pytype__ would be a type so that would work.

**Comment by bsteffensmeier** [comment]
> Do we need to remove Python 3.3 and 3.4 support? AppVeyor failed with those.

I would like to remove 3.3 and 3.4 support because it is easier than figuring out what is wrong and in my opinion anyone who needs new features should be upgrading their python.

**Comment by ndjensen** [comment]
Looks good to me.

**Pull request merged**